policy: Adjust existing policy for visibility annotations

jrajahalme opened the pull request:
Adjust and expand eBPF policy map keys and values to redirect for
visibility on the port of the visibility annotation while still
denying traffic on this port for identities for which the traffic is
denied.

Datapath lookup order is, from highest to lowest precedence:
1. L3/L4
2. L4-only (wildcard L3)
3. L3-only (wildcard L4)
4. Allow-all

This means that the L4-only allow visibility key can only be added if
there is an allow-all key, and all L3-only deny keys are expanded to
L3/L4 keys. If no L4-only key is added then also the L3-only allow
keys need to be expanded to L3/L4 keys for visibility redirection. In
addition the existing L3/L4 and L4-only allow keys need to be
redirected to the proxy port, if not already redirected.

The above can be accomplished by:

1. Change existing L4-only ALLOW key on matching port that does not already
   redirect to redirect.
   - e.g., 0:80=allow,0 -> 0:80=allow,<proxyport>
2. If allow-all key exists, add L4-only visibility redirect key if the L4-only
   key does not already exist.
   - e.g., 0:0=allow,0 -> add 0:80=allow,<proxyport> if 0:80 key does not exist
     - this allows all traffic on port 80, but see step 5 below.
3. Change all L3/L4 ALLOW keys on matching port that do not already redirect to
   redirect.
   - e.g, <ID1>:80=allow,0 -> <ID1>:80=allow,<proxyport>
4. For each L3-only ALLOW key add the corresponding L3/L4 ALLOW redirect if no
   L3/L4 key already exists and no L4-only key already exists and one is not added.
   - e.g., <ID2>:0=allow,0 -> add <ID2>:80=allow,<proxyport> if <ID2>:80
     and 0:80 do not exist and 0:80 was not added
5. If a new L4-only key was added: For each L3-only DENY key add the
   corresponding L3/L4 DENY key if no L3/L4 key already exists
   - e.g., <ID3>:0=deny,0 -> add <ID3>:80=deny,0 if <ID3>:80 does not exist.

With the above we only change/expand existing allow keys to redirect, and
expand existing drop keys to also drop on the port of interest, if a new
L4-only key allowing the port is added.

Adjust the unit test to mimic the real behavior a bit better by
regenerating endpoint policy after updating visibility policy. This is
needed due to the test using same port for HTTP and Kafka, and without
regenerating visibility redirects no longer overwrite existing
redirects.

Visibility redirects need to be re-applied after applying incremental
policy map changes. This is due to incremental changes possibly adding
keys for new identities that need visibility redirections or
additional deny keys as defined above.

Testing: New unit tests in policy package cover all the cases
described above.

Fixes: #9088

```release-note
Pod L7 visibility annotations are now supported also when policy enforcement is enabled.
```

jrajahalme committed:
policy: Adjust existing policy for visibility annotations

Adjust and expand eBPF policy map to redirect for visibility on the
port of the visibility annotation while still denying traffic on this
port for identities for which the traffic is denied.

Datapath lookup order is, from highest to lowest precedence:
1. L3/L4
2. L4-only (wildcard L3)
3. L3-only (wildcard L4)
4. Allow-all

This means that the L4-only allow visibility key can only be added if
there is an allow-all key, and all L3-only deny keys are expanded to
L3/L4 keys. If no L4-only key is added then also the L3-only allow
keys need to be expanded to L3/L4 keys for visibility redirection. In
addition the existing L3/L4 and L4-only allow keys need to be
redirected to the proxy port, if not already redirected.

The above can be accomplished by:

1. Change existing L4-only ALLOW key on matching port that does not already
   redirect to redirect.
   - e.g., 0:80=allow,0 -> 0:80=allow,<proxyport>
2. If allow-all key exists, add L4-only visibility redirect key if the L4-only
   key does not already exist.
   - e.g., 0:0=allow,0 -> add 0:80=allow,<proxyport> if 0:80 key does not exist
     - this allows all traffic on port 80, but see step 5 below
3. Change all L3/L4 ALLOW keys on matching port that do not already redirect to
   redirect.
   - e.g, <ID1>:80=allow,0 -> <ID1>:80=allow,<proxyport>
4. For each L3-only ALLOW key add the corresponding L3/L4 ALLOW redirect if no
   L3/L4 key already exists and no L4-only key already exists and one is not added.
   - e.g., <ID2>:0=allow,0 -> add <ID2>:80=allow,<proxyport> if <ID2>:80
     and 0:80 do not exist and 0:80 was not added
5. If a new L4-only key was added: For each L3-only DENY key add the
   corresponding L3/L4 DENY key if no L3/L4 key already exists
   - e.g., <ID3>:0=deny,0 -> add <ID3>:80=deny,0 if <ID3>:80 does not exist

With the above we only change/expand existing allow keys to redirect, and
expand existing drop keys to also drop on the port of interest, if a new
L4-only key allowing the port is added.

Adjust the unit test to mimic the real behavior a bit better by
regenerating endpoint policy after updating visibility policy. This is
needed due to the test using same port for HTTP and Kafka, and without
regenerating visibility redirects no longer overwrite existing
redirects.

Visibility redirects need to be re-applied after applying incremental
policy map changes. This is due to incremental changes possibly adding
keys for new identities that need visibility redirections or
additional deny keys as defined above.

Testing: New unit tests in policy package cover all the cases
described above.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aanm reviewed (approved):
LGTM, please double check if we need to update `Documentation/policy/visibility.rst`, specially the Limitations section

joestringer reviewed (changes requested):
What's the plan for testing this?

jrajahalme committed:
policy: Keep track of dependent (deny) entries

MapState entries that should be deleted together with some other
entry are "dependent entries". These are created for some deny
entries. Keep track of them, and delete them when the main entry is
deleted.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Generalize MapState owners

Allow other types in addition to CachedSelectors to own MapState
entries. This is needed to properly track entires added due to deny
rules so that incremental map updates can be performed correctly.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Do not expose internal maps to callers

MapState internal state could be modified by callers via internal
maps. Nil them on entries stored outside of the MapState itself.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joestringer commented on pkg/policy/mapstate_test.go:
```diff
@@ -983,3 +1276,441 @@ func (ds *PolicyTestSuite) TestMapState_AccumulateMapChanges(c *check.C) {
+		adds: MapState{},
+		deletes: MapState{
+			testIngressKey(235, 0, 0): testEntry(0, true),
+			HttpIngressKey(235):       testEntry(0, true), // XXX
```
What is `// XXX` here?

The `deletes` seem right to me.

jrajahalme replied:
Remnant from the phase when the delete was not happening yet and was commented out. Will remove the `// XXX`

jrajahalme replied:
Fixed.

joestringer commented on pkg/policy/mapstate_test.go:
```diff
@@ -983,3 +1276,441 @@ func (ds *PolicyTestSuite) TestMapState_AccumulateMapChanges(c *check.C) {
+			testIngressKey(235, 0, 0): testEntry(0, true),
+			testIngressKey(236, 0, 0): testEntry(0, true),
+		},
+		deletes: MapState{},
```
This test has no `visArgs` and doesn't continue the previous test, so I'm not sure why it's here inside `TestMapState_AccumulateMapChangesOnVisibilityKeys()`.

By itself, this one seems very similar to the test on line 886, and the following tests seem to follow a similar pattern as the tests following the one on line 886.

joestringer replied:
The tests seem right through to line 1628 but given the lack of visibility policy in them, they seem to just duplicate the tests through lines 886-983.

jrajahalme replied:
I forgot to add the `visArgs` after copy&paste. Will add or remove.

jrajahalme replied:
Added `visArgs` to the remaining test cases, which revealed a bug where dependents of an entry being updated could get lost (fixed).

joestringer commented on pkg/policy/mapstate_test.go:
```diff
@@ -983,3 +1276,441 @@ func (ds *PolicyTestSuite) TestMapState_AccumulateMapChanges(c *check.C) {
+		deletes: MapState{},
+	}, {
+		continued: true,
+		name:      "test-2 - Removing the sole key",
```
Earlier, when a test continues, you used the naming scheme `test-N-[a-z]`, but here it's continuing with a different test number. Not a big deal but would be nice to be consistent.

jrajahalme replied:
Will fix, `a-z` for continued tests makes sense.

jrajahalme replied:
Fixed.

joestringer commented on pkg/policy/mapstate_test.go:
```diff
@@ -983,3 +1276,441 @@ func (ds *PolicyTestSuite) TestMapState_AccumulateMapChanges(c *check.C) {
+		deletes: MapState{},
+	}, {
+		continued: false,
+		name:      "test-7a - egress HTTP proxy (setup)",
```
Now the tests seem to be copying those from line 1198 onwards :thinking: Am I missing some difference here or was visibility policy intended to be introduced into the ones in this function?

jrajahalme replied:
Not sure any more, will check :-)

jrajahalme replied:
Added visibility redirects to this test as well.

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -56,6 +56,9 @@ type CachedSelector interface {
 	String() string
+
+	// RemoveDependent implements MapStateOwner interface
+	RemoveDependent(keys MapState, key Key)
```
I had a bit of trouble trying to understand the notion of the dependent entries based purely on the commit message and code introduced in that third patch. I think both reviewers and future Cilium contributors would really appreciate a single full example written up somewhere to give a better sense for the relationship that dependents have with their parent entries. I guess we can figure it out from the tests but it is a bit terse.

At least one take on what a "dependent" is would be: A rule derived from two sources: Firstly, a visibility policy which should redirect this traffic towards a proxy; and secondly, a policy allow which selects the traffic that the visibility policy applies to, which is the basis for allowing that visibility policy to be installed. So, what this means is that if a visibility rule (MapStateEntry) is dependent on some broader allow rule, and then the broader allow rule is removed, then we will also remove the visibility rule (because the visibility policy doesn't determine allow policy all by itself).

Maybe the above is a useful startpoint to put a description somewhere? I think there's more complicated cases though as well, from the tests I saw that there are additional constraints around what happens when there's a combination of allows and denies plus a visibility policy. So if you have an allow rule, a deny rule, and a visibility policy which may select traffic that is selected by both the allow and the deny rule... then there may be dependent rules both for the visibility policy itself but also for the deny policies to ensure that the deny still takes precedence over the visibility + allow rules.

joestringer replied:
Having got to patch 4/4 of review now, I see that at least for the visibility cases it is pretty well-documented, :pray: . In my descriptions above I assumed that dependent entries are only used with visibility policy, but evidently in patch 3 they're used when you combine allow-all and L4-only allows with L3+L4 denies.

joestringer commented on pkg/policy/mapstate_test.go:
```diff
@@ -727,57 +739,338 @@ func (ds *PolicyTestSuite) TestMapState_DenyPreferredInsert(c *check.C) {
+		},
+		state: MapState{
+			AnyIngressKey():          testEntry(0, false),
+			testIngressKey(41, 0, 0): testEntry(0, true, csFoo).WithDependents(HttpIngressKey(41)),
```
I didn't see any test cases with multiple dependents, are there any tricky corner cases there?

jrajahalme replied:
Added a test case and found that the deny code was adding modified entries without clearing the dependents. Changed to always create new dependent entries with `NewMapStateEntry()`.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -195,29 +255,46 @@ func (keys MapState) addKeyWithChanges(key Key, entry MapStateEntry, adds, delet
-				// key is not deleted if other selectors still need it
-				if len(entry.selectors) > 0 {
+				delete(entry.owners, owner)
+				owner.RemoveDependent(keys, key)
```
I'm curious why it makes sense to add the `RemoveDependent()` abstraction on the `MapStateOwner` for this implementation rather than against the `MapState`? I understand that the `owner` likely needs to be abstracted to be received into this function, but it's a bit odd that `MapStateOwner` interface only requires `RemoveDependent()` to be implemented (and not `AddDependent()`), and also that half of the implementers implement the function as a no-op (`CachedSelector`).

jrajahalme replied:
The call sites will need to discriminate on the actual type of the `MapStateOwner` interface, but I agree, it will be clearer. Initially I thought that I could have `RemoveDependent()` only take a key, but now it also needs to take the map so it is better to make this a function on the map.

joestringer commented on pkg/policy/mapstate_test.go:
```diff
@@ -727,57 +739,338 @@ func (ds *PolicyTestSuite) TestMapState_DenyPreferredInsert(c *check.C) {
-	HttpIngressKey := func(id int) Key {
-		return TestIngressKey(id, 80, 6)
+	return entry
+}
```
For what it's worth, I found this easier to read by adding these small changes:

```
diff --git a/pkg/policy/mapstate_test.go b/pkg/policy/mapstate_test.go
index ab6b0e3cb41d..b382e847d3bd 100644
--- a/pkg/policy/mapstate_test.go
+++ b/pkg/policy/mapstate_test.go
@@ -798,6 +798,14 @@ func testEntry(proxyPort uint16, deny bool, owners ...MapStateOwner) MapStateEnt
        return entry
 }
 
+func allowEntry(proxyPort uint16, owners ...MapStateOwner) MapStateEntry {
+       return testEntry(proxyport, false, owners...)
+}
+
+func denyEntry(proxyPort uint16, owners ...MapStateOwner) MapStateEntry {
+       return testEntry(proxyport, true, owners...)
+}
+
 func testEntryD(proxyPort uint16, deny bool, derivedFrom labels.LabelArrayList, owners ...MapStateOwner) MapStateEntry {
        entry := testEntry(proxyPort, deny, owners...)
        entry.DerivedFromRules = derivedFrom
```

....and then doing a couple of sed expressions on the rest of the file so that I could read each test as a series of `allowEntry` / `denyEntry` callouts.

If you think this makes it easier, I'd be in favour of adding this change in, but not required :)

jrajahalme replied:
Good point, will adopt :-)

jrajahalme replied:
Done.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -195,29 +255,46 @@ func (keys MapState) addKeyWithChanges(key Key, entry MapStateEntry, adds, delet
 			}
 		}
+
+		// Remove this key from all owners' dependents maps if no owner was given
```
This is the case this is handling, right? Would it help provide better context to readers if we added this small extra comment here?

```suggestion
		// Remove this key from all owners' dependents maps if no owner was given
		// This is typically the case if the key is being removed due to a deny rule being added.
```

jrajahalme replied:
Actually I'm not sure if dependent rules are ever deleted without passing the owner, added this for completeness so that if that ever happens there will be do dangling references left behind.

joestringer replied:
When I searched for the caller of this function, I found some calls which explicitly passed `nil` as the `MapStateOwner` parameter so I think this is actually being used. That's where I came up with the comment suggestion.

jrajahalme replied:
Added this to the comment here:
```
		// Owner is nil when deleting more specific entries (e.g., L3/L4) when
		// adding deny entries that cover them (e.g., L3-deny).
```

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -241,13 +318,20 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
 				if newKey.TrafficDirection == k.TrafficDirection &&
 					!v.IsDeny &&
-					k.Identity == 0 {
+					k.Identity == 0 && (k.DestPort != 0 || k.Nexthdr != 0) {
```
What happens if there's an allow-all entry already and now we're introducing an l3-only deny?

jrajahalme replied:
As per the datapath lookup order nothing special is needed, as the L3-only deny will be looked up before the allow-all. Deny rules always take precedence so this is the correct thing to do.

joestringer replied:
Ah okay. At some level that makes sense. I think I was just caught up a bit by the fact that this condition was changed, so I was wondering if we were just running extra unnecessary code in this snippet before for that case or we could have limited this condition to L4-only entries from the beginning.

Like is this change by itself safe regardless of the rest of the PR? Or is it depending on some other logic being changed somewhere else?

jrajahalme replied:
Added a test case to cover this explicitly (`test-4 - Adding L3-deny to an existing allow-all`)

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -314,25 +406,190 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+// expand existing drop keys to also drop on the port of interest, if a new
+// L4-only key allowing the port is added.
+//
+// 'adds' and 'deletes' are updates with the changes made. 'adds' contains the new values
```
```suggestion
// 'adds' and 'deletes' are updated with the changes made. 'adds' contains the new values
```

Any particular reason to pass in these output parameters rather than just returning them? Is it just to avoid the merging logic in the caller?

jrajahalme replied:
Yes. All these changes are collected into `MapStateChanges` so it is logical to pass them around as these maps exist independent of any one call being made.

jrajahalme replied:
Fixed typo.

joestringer commented on pkg/policy/mapstate_test.go:
```diff
@@ -727,57 +739,338 @@ func (ds *PolicyTestSuite) TestMapState_DenyPreferredInsert(c *check.C) {
+		adds      MapState
+		deletes   MapState
+	}{{
+		name: "test-1a - Adding identity to an existing state",
```
nit:
```suggestion
		name: "test-1a - Add L3+L4 deny to existing Allow-all, L4-only allow rule map state entries",
```

jrajahalme replied:
+1

jrajahalme replied:
Changed title to `test-1a - Adding L3-deny to an existing allow-all with L4-only allow redirect map state entries` - the L3/L4 deny is then added due to the L4-allow still being in effect for all other L3's.

joestringer reviewed (commented):
Core implementation looks good. There was a couple of corner cases I didn't see tests for, and a few tests I was a little bit confused by. The rest is a scattering of misc comments while I figured stuff out.

joestringer reviewed (approved):
Temporarily approving in order to test something for GitHub support.

joestringer reviewed (changes requested):
My previous round of review still applies.

joestringer commented on pkg/policy/mapstate_test.go:
```diff
@@ -1497,153 +1553,235 @@ func (ds *PolicyTestSuite) TestMapState_AccumulateMapChangesOnVisibilityKeys(c *
+			AnyIngressKey():  allowEntry(0),
+			HostIngressKey(): allowEntry(0),
+			// Entry added solely due to visibility annotation has a 'nil' owner
+			HttpIngressKey(0): allowEntryD(12345, visibilityDerivedFrom).WithOwners(nil),
```
This test seems a little inconsistent with respect to the ingress visibility policy. In the `setup()`, there's an ingress allow-all and ingress allow-host but no L4 ingress keys. Now this line demonstrates that the resulting state includes an HTTP ingress L4 entry, which I would expect given the visibility policy on redirect port 12345. So this is state that changes when running the logic under test, yet it doesn't end up in the "adds" below.

I understand that this test is only testing the "add" of a couple of egress L3+L4 entries, but if that's the case I'm a bit confused about the role of the ingress visibility policy in this test. Is it to deliberately check that when you add egress allow rules, it will only affect egress and won't touch the ingress rules? But if that's the case, why does the 'state' change to include the ingress visibility policy?

jrajahalme replied:
The test applies `visArgs` twice, first before applying `args` and then after applying `args`. The `adds` and `deletes` in the test data only account for the latter with the intent of capturing the changes due to `args` only. We could eliminate the first pass of `visArgs` by requiring that `state` already include those keys. First I thought that this would be too hard on the test writer (to manually compute the entries added due to `visArgs`), but now I think I should do that to remove this potential for confusion & but keep the 1st pass of `visArgs` to verify there are no changes (i.e., `state` is consistent).

jrajahalme replied:
Ended up adding `visAdds` and `visDeletes` to verify the keys added and deleted on the first round of applying `visArgs`. This verifies that those changes are made before applying `args`.

Will squash these changes to the main commit, please review this unit test again, there are no changes elsewhere :-)

joestringer reviewed (commented):
New changes LGTM, just one test I'm a bit confused by (thread below).

joestringer reviewed (approved):
Latest changes LGTM.
